core/domain/classifier/base_classification_algorithm.py

```diff
@@ -0,0 +1,153 @@
+# coding: utf-8
+#
+# Copyright 2015 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base class of classification algorithms for free-form text answers"""
+
+from abc import abstractmethod
+import copy
+
+
+class BaseClassificationAlgorithm(object):
+    """A base class for classifiers that uses supervised learning to match
```

Comment:
In the long term, we are looking to have classifiers for code also. This is something which we don't have to worry about now but our design decisions should take into account the fact that we are not just dealing with text. So at this point what we really have to think about is a structure which can handle multiple algorithms for text classification & multiple algorithms for program (code) classification. So the real hierarchy (just to make things clear) is like: Classifier->(Text Classifier -> (multiple text classifiers), Code Classifier -> (multiple code classifiers)). 

Now, do we really need a separate layer for abstracting the family of string and code classifiers? Well, that's something we need to dig into & see what are the pros/cons of providing the abstraction vs not providing the abstraction. We need to think over this and put our proposal in the design document. I think this is a good design problem.

Also, we need to clearly describe in the design doc about how we will decide what algorithm will be used for classification (the creator shouldn't decide here because this entire process is hidden from him). Secondly, we also need to describe how we will work with models that have been generated using different classification algorithms.

@seanlip would you like to add anything?

PS: @Andimeo If you want a background information on code classification, please take a look [here](https://github.com/oppia/oppia/wiki/Google-Summer-of-Code-2016#applying-machine-learning-to-code-interaction-training-and-generalizing).